File: Alien Dictionary.java
class Solution {
    public String findOrder(String[] words) {
        HashMap<Character, HashSet<Character>> adj = new HashMap<>();
        HashMap<Character, Integer> inDeg = new HashMap<>();
        for(String word : words){
            for(int i = 0; i < word.length(); i++){
                inDeg.put(word.charAt(i), 0);
            }
        }
        int totalChars = inDeg.size();
        
        for(int i = 1; i < words.length; i++){
            String past = words[i - 1];
            String curr = words[i];
            int l1 = 0, l2 = 0;
            while(l1 < past.length() && l2 < curr.length()){
                if(past.charAt(l1) != curr.charAt(l2)){
                    if(adj.containsKey(past.charAt(l1))){
                        if(!adj.get(past.charAt(l1)).contains(curr.charAt(l2))) inDeg.put(curr.charAt(l2), inDeg.get(curr.charAt(l2)) + 1);
                        adj.get(past.charAt(l1)).add(curr.charAt(l2));
                    }
                    else{
                        HashSet<Character> hs = new HashSet<>();
                        hs.add(curr.charAt(l2));
                        adj.put(past.charAt(l1), hs);
                        inDeg.put(curr.charAt(l2), inDeg.get(curr.charAt(l2)) + 1);
                    }
                    break;
                }
                l1++;
                l2++;
            }
            if(l1 < past.length() && l2 == curr.length()) return "";
        }
        
        Queue<Character> q = new LinkedList<>();
        for(Map.Entry<Character, Integer> ele : inDeg.entrySet()){
            if(ele.getValue() == 0) q.offer(ele.getKey());
        }
        
        StringBuilder sb = new StringBuilder();
        while(!q.isEmpty()){
            char ch = q.poll();
            sb.append(ch);
            
            if(!adj.containsKey(ch)) continue;
            for(Character neig : adj.get(ch)){
                inDeg.put(neig, inDeg.get(neig) - 1);
                if(inDeg.get(neig) == 0){
                    q.offer(neig);
                }
            }
        }
        
        if(sb.length() != inDeg.size()) return "";
        return sb.toString();
    }
}

File: Alien Dictionary.cpp
class Solution {
public:
    std::string findOrder(std::vector<std::string>& words) {
        std::unordered_map<char, std::unordered_set<char>> adj;
        std::unordered_map<char, int> inDeg;
        for (const auto& word : words) {
            for (char c : word) {
                inDeg[c] = 0;
            }
        }
        int totalChars = inDeg.size();

        for (size_t i = 1; i < words.size(); i++) {
            const std::string& past = words[i - 1];
            const std::string& curr = words[i];
            size_t l1 = 0, l2 = 0;
            while (l1 < past.size() && l2 < curr.size()) {
                if (past[l1] != curr[l2]) {
                    if (adj.count(past[l1])) {
                        if (!adj[past[l1]].count(curr[l2])) {
                            inDeg[curr[l2]]++;
                        }
                        adj[past[l1]].insert(curr[l2]);
                    } else {
                        adj[past[l1]] = {curr[l2]};
                        inDeg[curr[l2]]++;
                    }
                    break;
                }
                l1++;
                l2++;
            }
            if (l1 < past.size() && l2 == curr.size()) return "";
        }

        std::queue<char> q;
        for (const auto& ele : inDeg) {
            if (ele.second == 0) q.push(ele.first);
        }

        std::string sb;
        while (!q.empty()) {
            char ch = q.front();
            q.pop();
            sb.push_back(ch);

            if (!adj.count(ch)) continue;
            for (char neig : adj[ch]) {
                inDeg[neig]--;
                if (inDeg[neig] == 0) {
                    q.push(neig);
                }
            }
        }

        if (sb.size() != inDeg.size()) return "";
        return sb;
    }
};